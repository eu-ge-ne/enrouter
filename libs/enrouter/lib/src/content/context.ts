import { createContext, useContext } from "react";

import type { Content } from "./mod.js";

const ContentContext = createContext<Content>({
  match: {
    route: {
      path: "",
      test: {
        keys: [],
        pattern: new RegExp(""),
      },
      modules: [],
      loaded: false,
      elements: {},
    },
    location: "",
    isFull: false,
    params: {},
  },
});

export const ContentProvider = ContentContext.Provider;

export function useContent(): Content {
  return useContext(ContentContext);
}
