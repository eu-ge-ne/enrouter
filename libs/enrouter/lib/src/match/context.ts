import { createContext, useContext } from "react";

import type { Match } from "./mod.js";

const MatchContext = createContext<Match>({
  route: {
    path: "",
    test: { keys: [], pattern: new RegExp("") },
    modules: [],
    loaded: false,
    elements: {},
  },

  isFull: false,
  location: "",
  params: {},
});

export const MatchProvider = MatchContext.Provider;

export function useMatch(): Match {
  return useContext(MatchContext);
}
