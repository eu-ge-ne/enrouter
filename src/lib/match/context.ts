import { createContext, useContext } from "react";

import type { Match } from "./match.js";

const MatchesContext = createContext<Match[]>([]);
const MatchIndexContext = createContext<number>(-1);

export const MatchesProvider = MatchesContext.Provider;
export const MatchIndexProvider = MatchIndexContext.Provider;

export interface MatchContext {
  matches: Match[];
  matchIndex: number;
  match: Match | undefined;
  firstMatch: Match | undefined;
  nextMatch: Match | undefined;
  lastMatch: Match | undefined;
  voidMatch: Match | undefined;
  isExactMatch: boolean;
}

export function useMatchContext(): MatchContext {
  const matches = useContext(MatchesContext);
  const matchIndex = useContext(MatchIndexContext);

  const match = matches[matchIndex];
  const firstMatch = matches[0];
  const nextMatch = matches[matchIndex + 1];
  const lastMatch = matches.at(-1);
  const voidMatch = matches.findLast((x) => x.route?.elements._void);

  const isExactMatch = Boolean(lastMatch?.route);

  return {
    matches,
    matchIndex,
    match,
    firstMatch,
    nextMatch,
    lastMatch,
    voidMatch,
    isExactMatch,
  };
}

export function useMatch(path: string): Match | undefined {
  return useContext(MatchesContext).find((x) => x.route?.path === path);
}
