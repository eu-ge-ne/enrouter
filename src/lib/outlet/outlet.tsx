import type { ReactNode, ReactElement } from "react";

import type { Match } from "#lib/match/match.js";
import {
  MatchIndexProvider,
  useMatches,
  useMatchIndex,
} from "#lib/match/context.js";

export interface OutletProps {
  name?: string;
}

export function Outlet({ name }: OutletProps): ReactNode {
  const matches = useMatches();
  const matchIndex = useMatchIndex();

  // root?
  if (matchIndex < 0) {
    // void

    // next?
    if (matches[0]) {
      return <Next index={0} match={matches[0]} name={name} />;
    }

    return;
  }

  const match = matches[matchIndex]!;
  const nextMatch = matches[matchIndex + 1];
  const lastMatch = matches.at(-1)!;

  // void?
  if (!lastMatch.isExact) {
    const lastVoid = matches.findLast((x) => x.route.elements._void);
    if (match === lastVoid) {
      return pick(match.route.elements._void, name);
    }
  }

  // content?
  if (match.isExact && match === lastMatch) {
    return pick(match.route.elements._content, name);
  }

  // next?
  if (nextMatch) {
    return <Next index={matchIndex + 1} match={nextMatch} name={name} />;
  }
}

interface NextProps {
  index: number;
  match: Match;
  name: string | undefined;
}

function Next({ index, match, name }: NextProps): ReactElement {
  const { _layout, _content } = match.route.elements;

  return (
    <MatchIndexProvider value={index}>
      {pick(_layout ?? _content, name)}
    </MatchIndexProvider>
  );
}

function pick(
  els: Record<string, ReactElement> | undefined,
  name?: string
): ReactNode {
  if (els) {
    return name ? els[name] : Object.values(els)[0];
  }
}
