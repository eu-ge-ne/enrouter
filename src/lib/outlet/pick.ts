import type { ReactNode, ReactElement } from "react";

export function pick(
  els: Record<string, ReactElement> | undefined,
  name?: string,
): ReactNode {
  if (els) {
    return name ? els[name] : Object.values(els)[0];
  }
}
