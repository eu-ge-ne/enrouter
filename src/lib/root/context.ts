import { type ComponentType, createContext, useContext } from "react";

export interface RootParams {
  root?: ComponentType;
  fallback?: Record<string, ComponentType>;
}

const RootParamsContext = createContext<RootParams>({});

export const RootParamsProvider = RootParamsContext.Provider;

export function useRootParams(): RootParams {
  return useContext(RootParamsContext);
}
