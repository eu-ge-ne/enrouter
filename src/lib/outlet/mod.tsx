import type { ReactNode, ReactElement } from "react";

import { MatchProvider, useMatch } from "#lib/match/context.js";

export interface OutletProps {
  name?: string;
}

export function Outlet({ name }: OutletProps): ReactNode {
  const match = useMatch();
  if (!match) {
    return;
  }

  if (!match.last?.isFull) {
    if (match.isVoid) {
      return pick(match.route.elements._void, name);
    }
  }

  if (match.next) {
    const { _layout, _content } = match.next.route.elements;

    return (
      <MatchProvider value={match.next}>
        {pick(_layout ?? _content, name)}
      </MatchProvider>
    );
  }

  return pick(match.route.elements._content, name);
}

function pick(
  els: Record<string, ReactElement> | undefined,
  name?: string,
): ReactElement | undefined {
  if (els) {
    return name ? els[name] : Object.values(els)[0];
  }
}
