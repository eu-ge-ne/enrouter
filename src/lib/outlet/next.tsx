import type { ReactNode, ReactElement } from "react";

import type { Match } from "#lib/match/match.js";
import { MatchIndexProvider } from "#lib/match/context.js";
import { pick } from "./pick.js";

interface NextProps {
  index: number;
  match: Match;
  name: string | undefined;
}

export function Next({ index, match, name }: NextProps): ReactElement {
  const { _layout, _content } = match.route.elements;

  return (
    <MatchIndexProvider value={index}>
      {pick(_layout ?? _content, name)}
    </MatchIndexProvider>
  );
}
