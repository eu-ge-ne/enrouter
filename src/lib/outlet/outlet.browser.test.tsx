import type { FC, PropsWithChildren } from "react";
import { describe, test, expect } from "vitest";
import { render } from "vitest-browser-react";
import * as regexparam from "regexparam";

import type { Match } from "#lib/match/match.js";
import { MatchesProvider, MatchIndexProvider } from "#lib/match/context.js";
import { type RootParams, RootParamsProvider } from "#lib/root/context.js";
import { Outlet } from "./outlet.js";

const testId = "test-wrapper";

const wrapper: FC<PropsWithChildren> = ({ children }) => (
  <div data-testid={testId}>{children}</div>
);

describe("outlet", () => {
  test("no match", async () => {
    const screen = render(
      <MatchesProvider value={[]}>
        <Outlet />
      </MatchesProvider>,
      { wrapper },
    );

    await expect.element(screen.getByTestId(testId)).toBeVisible();

    expect(screen.container).toMatchSnapshot();
  });

  test("_layout from next match", async () => {
    const matches: Match[] = [
      {
        route: {
          path: "/",
          test: regexparam.parse("/", true),
          modules: [],
          loaded: true,
          elements: {},
        },
        location: "/",
        params: {},
      },
      {
        route: {
          path: "/abc",
          test: regexparam.parse("/abc", true),
          modules: [],
          loaded: true,
          elements: {
            _layout: {
              Main: <div>next layout#Main</div>,
            },
          },
        },
        location: "/abc",
        params: {},
      },
    ];

    const screen = render(
      <MatchesProvider value={matches}>
        <MatchIndexProvider value={0}>
          <Outlet name="Main" />
        </MatchIndexProvider>
      </MatchesProvider>,
      { wrapper },
    );

    await expect.element(screen.getByTestId(testId)).toBeVisible();

    expect(screen.container).toMatchSnapshot();
  });

  test("_content from next match", async () => {
    const matches: Match[] = [
      {
        route: {
          path: "/",
          test: regexparam.parse("/", true),
          modules: [],
          loaded: true,
          elements: {},
        },
        location: "/",
        params: {},
      },
      {
        route: {
          path: "/abc",
          test: regexparam.parse("/abc", true),
          modules: [],
          loaded: true,
          elements: {
            _content: {
              Main: <div>next content#Main</div>,
            },
          },
        },
        location: "/abc",
        params: {},
      },
    ];

    const screen = render(
      <MatchesProvider value={matches}>
        <MatchIndexProvider value={0}>
          <Outlet name="Main" />
        </MatchIndexProvider>
      </MatchesProvider>,
      { wrapper },
    );

    await expect.element(screen.getByTestId(testId)).toBeVisible();

    expect(screen.container).toMatchSnapshot();
  });

  test("_content from current match", async () => {
    const matches: Match[] = [
      {
        route: {
          path: "/",
          test: regexparam.parse("/", true),
          modules: [],
          loaded: true,
          elements: {
            _content: {
              Main: <div>content#Main</div>,
            },
          },
        },
        location: "/",
        params: {},
      },
    ];

    const screen = render(
      <MatchesProvider value={matches}>
        <MatchIndexProvider value={0}>
          <Outlet name="Main" />
        </MatchIndexProvider>
      </MatchesProvider>,
      { wrapper },
    );

    await expect.element(screen.getByTestId(testId)).toBeVisible();

    expect(screen.container).toMatchSnapshot();
  });

  test("_fallback from current match", async () => {
    const matches: Match[] = [
      {
        route: {
          path: "/",
          test: regexparam.parse("/", true),
          modules: [],
          loaded: true,
          elements: {
            _fallback: {
              Main: <div>fallback#Main</div>,
            },
          },
        },
        location: "/",
        params: {},
      },
      {
        location: "/abc",
        params: {},
      },
    ];

    const screen = render(
      <MatchesProvider value={matches}>
        <MatchIndexProvider value={0}>
          <Outlet name="Main" />
        </MatchIndexProvider>
      </MatchesProvider>,
      { wrapper },
    );

    await expect.element(screen.getByTestId(testId)).toBeVisible();

    expect(screen.container).toMatchSnapshot();
  });

  test("_layout from first match", async () => {
    const matches: Match[] = [
      {
        route: {
          path: "/",
          test: regexparam.parse("/", true),
          modules: [],
          loaded: true,
          elements: {
            _layout: {
              Main: <div>layout#Main</div>,
            },
          },
        },
        location: "/",
        params: {},
      },
    ];

    const screen = render(
      <MatchesProvider value={matches}>
        <Outlet />
      </MatchesProvider>,
      { wrapper },
    );

    await expect.element(screen.getByTestId(testId)).toBeVisible();

    expect(screen.container).toMatchSnapshot();
  });

  test("_content from first match", async () => {
    const matches: Match[] = [
      {
        route: {
          path: "/",
          test: regexparam.parse("/", true),
          modules: [],
          loaded: true,
          elements: {
            _content: {
              Main: <div>content#Main</div>,
            },
          },
        },
        location: "/",
        params: {},
      },
    ];

    const screen = render(
      <MatchesProvider value={matches}>
        <Outlet />
      </MatchesProvider>,
      { wrapper },
    );

    await expect.element(screen.getByTestId(testId)).toBeVisible();

    expect(screen.container).toMatchSnapshot();
  });

  test("root fallback", async () => {
    const rootParams: RootParams = {
      fallback: {
        X: () => <div>fallback</div>,
      },
    };

    const matches: Match[] = [
      {
        route: {
          path: "/",
          test: regexparam.parse("/", true),
          modules: [],
          loaded: true,
          elements: {},
        },
        location: "/",
        params: {},
      },
      {
        location: "/x",
        params: {},
      },
    ];

    const screen = render(
      <RootParamsProvider value={rootParams}>
        <MatchesProvider value={matches}>
          <Outlet />
        </MatchesProvider>
      </RootParamsProvider>,
      { wrapper },
    );

    await expect.element(screen.getByTestId(testId)).toBeVisible();

    expect(screen.container).toMatchSnapshot();
  });

  test("named root fallback", async () => {
    const rootParams: RootParams = {
      fallback: {
        A: () => <div>fallback A</div>,
      },
    };

    const matches: Match[] = [
      {
        route: {
          path: "/",
          test: regexparam.parse("/", true),
          modules: [],
          loaded: true,
          elements: {},
        },
        location: "/",
        params: {},
      },
      {
        location: "/x",
        params: {},
      },
    ];

    const screen = render(
      <RootParamsProvider value={rootParams}>
        <MatchesProvider value={matches}>
          <Outlet name="A" />
        </MatchesProvider>
      </RootParamsProvider>,
      { wrapper },
    );

    await expect.element(screen.getByTestId(testId)).toBeVisible();

    expect(screen.container).toMatchSnapshot();
  });
});
