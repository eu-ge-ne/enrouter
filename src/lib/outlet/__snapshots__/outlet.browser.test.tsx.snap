// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`outlet > _content from current match 1`] = `
<div>
  <div
    data-testid="test-wrapper"
  >
    <div>
      content#Main
    </div>
  </div>
</div>
`;

exports[`outlet > _content from first match 1`] = `
<div>
  <div
    data-testid="test-wrapper"
  >
    <div>
      content#Main
    </div>
  </div>
</div>
`;

exports[`outlet > _content from next match 1`] = `
<div>
  <div
    data-testid="test-wrapper"
  >
    <div>
      next content#Main
    </div>
  </div>
</div>
`;

exports[`outlet > _layout from first match 1`] = `
<div>
  <div
    data-testid="test-wrapper"
  >
    <div>
      layout#Main
    </div>
  </div>
</div>
`;

exports[`outlet > _layout from next match 1`] = `
<div>
  <div
    data-testid="test-wrapper"
  >
    <div>
      next layout#Main
    </div>
  </div>
</div>
`;

exports[`outlet > _void from current match 1`] = `
<div>
  <div
    data-testid="test-wrapper"
  >
    <div>
      void#Main
    </div>
  </div>
</div>
`;

exports[`outlet > named root void 1`] = `
<div>
  <div
    data-testid="test-wrapper"
  >
    <div>
      void A
    </div>
  </div>
</div>
`;

exports[`outlet > no match 1`] = `
<div>
  <div
    data-testid="test-wrapper"
  />
</div>
`;

exports[`outlet > root void 1`] = `
<div>
  <div
    data-testid="test-wrapper"
  >
    <div>
      void
    </div>
  </div>
</div>
`;
